azure-mgmt-cdn/src/main/java/com/microsoft/azure/management/cdn/implementation/CdnEndpointsImpl.java

```diff
@@ -0,0 +1,154 @@
+import com.microsoft.azure.management.cdn.CdnProfile;
+import com.microsoft.azure.management.cdn.CheckNameAvailabilityResult;
+import com.microsoft.azure.management.cdn.DeepCreatedOrigin;
+import com.microsoft.azure.management.resources.fluentcore.arm.collection.implementation.ExternalChildResourcesCachedImpl;
+import com.microsoft.azure.management.resources.fluentcore.utils.ResourceNamer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents an endpoint collection associated with a CDN manager profile.
+ */
+class CdnEndpointsImpl extends
+        ExternalChildResourcesCachedImpl<CdnEndpointImpl,
+                        CdnEndpoint,
+                        EndpointInner,
+                        CdnProfileImpl,
+                        CdnProfile> {
+    private final EndpointsInner client;
+    private final OriginsInner originsClient;
+    private final CustomDomainsInner customDomainsClient;
+
+    CdnEndpointsImpl(EndpointsInner client,
+                     OriginsInner originsClient,
+                     CustomDomainsInner customDomainsClient,
+                     CdnProfileImpl parent) {
+        super(parent, "Endpoint");
+        this.client = client;
+        this.originsClient = originsClient;
+        this.customDomainsClient = customDomainsClient;
+    }
+
+    /**
+     * @return the azure endpoints as a map indexed by name.
+     */
+    Map<String, CdnEndpoint> endpointsAsMap() {
+        Map<String, CdnEndpoint> result = new HashMap<>();
+        for (Map.Entry<String, CdnEndpointImpl> entry : this.collection().entrySet()) {
+            CdnEndpointImpl endpoint = entry.getValue();
+            result.put(entry.getKey(), endpoint);
+        }
+        return Collections.unmodifiableMap(result);
+    }
+
+    /**
+     * Mark the endpoint with given name as to be removed.
+     *
+     * @param name the name of the endpoint to be removed
+     */
+    public void remove(String name) {
+        this.prepareRemove(name);
+    }
+
+    /**
+     * Adds the endpoint to the collection.
+     *
+     * @param endpoint the endpoint
+     */
+    public void addEndpoint(CdnEndpointImpl endpoint) {
+        this.addChildResource(endpoint);
+    }
+
+    @Override
+    protected List<CdnEndpointImpl> listChildResources() {
+        List<CdnEndpointImpl> childResources = new ArrayList<>();
+
+        for (EndpointInner innerEndpoint : this.client.listByProfile(
+                                        this.parent().resourceGroupName(),
+                                        this.parent().name())) {
+            childResources.add(new CdnEndpointImpl(innerEndpoint.name(),
+                    this.parent(),
+                    innerEndpoint,
+                    this.client,
+                    this.originsClient,
+                    this.customDomainsClient));
+        }
+        return childResources;
```

Comment:
is this list ultimately returned to the user code? If so, shouldn't it go through Collections.unmodifiableList?